database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/PostViewModel.swift

```diff
@@ -0,0 +1,123 @@
+//
+//  Copyright (c) 2021 Google Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import SwiftUI
+import Firebase
+
+class PostViewModel: ObservableObject, Identifiable {
+  @Published var id: String
+  @Published var uid: String
+  @Published var author: String
+  @Published var title: String
+  @Published var body: String
+  @Published var starCount: Int
+  @Published var userIDsStarredBy: [String: Bool]
+  var postRef: DatabaseReference!
+
+  init(id: String, uid: String, author: String, title: String, body: String) {
+    self.id = id
+    self.uid = uid
+    self.author = author
+    self.title = title
+    self.body = body
+    starCount = 0
+    userIDsStarredBy = [:]
+  }
+
+  init?(id: String, dict: [String: Any]) {
+    guard let uid = dict["uid"] as? String else { return nil }
+    guard let author = dict["author"] as? String else { return nil }
+    guard let title = dict["title"] as? String else { return nil }
+    guard let body = dict["body"] as? String else { return nil }
+    let userIDsStarredBy = dict["userIDsStarredBy"] as? [String: Bool] ?? [:]
+    let starCount = dict["starCount"] as? Int ?? 0
+
+    self.id = id
+    self.uid = uid
+    self.author = author
+    self.title = title
+    self.body = body
+    self.starCount = starCount
+    self.userIDsStarredBy = userIDsStarredBy
+  }
+
+  func isStarred() -> Bool {
+    if let uid = Auth.auth().currentUser?.uid {
+      return userIDsStarredBy[uid] ?? false
+    }
+    return false
+  }
+
+  func didTapStarButton() {
+    // updating local values
+    if let uid = Auth.auth().currentUser?.uid {
+      if let _ = userIDsStarredBy[uid] {
+        starCount -= 1
+        userIDsStarredBy.removeValue(forKey: uid)
+      } else {
+        starCount += 1
+        userIDsStarredBy[uid] = true
+      }
+    }
+
+    // updating firebase values
+    postRef = Database.database().reference().child("posts").child(id)
+    incrementStars(forRef: postRef)
+    postRef.observeSingleEvent(of: .value, with: { snapshot in
+      guard let value = snapshot.value as? [String: Any] else { return }
+      if let uid = value["uid"] as? String {
+        let userPostRef = Database.database().reference()
+          .child("user-posts")
+          .child(uid)
+          .child(self.id)
+        self.incrementStars(forRef: userPostRef)
+      }
+    })
+  }
+
+  func incrementStars(forRef ref: DatabaseReference) {
+    // [START post_stars_transaction]
```

Comment:
These START and END tags are for connecting to code snippets from the docs. Have you coordinated with @morganchen12 about the implications?

Reply:
No START or END tags are necessary in the new samples, please just assign any includecode bugs to me and I'll update the paths to point to the new locations of the old samples.